File: src/shell/states/Binary.zig
pub const Binary = @This();

base: State,
node: *const ast.Binary,
/// Based on precedence rules binary expr can only be child of a stmt or
/// another binary expr
parent: ParentPtr,
left: ?ExitCode = null,
right: ?ExitCode = null,
io: IO,
currently_executing: ?ChildPtr = null,

pub const ChildPtr = StatePtrUnion(.{
    Async,
    Cmd,
    Pipeline,
    Binary,
    Assigns,
    If,
    CondExpr,
    Subshell,
});

pub const ParentPtr = StatePtrUnion(.{
    Stmt,
    Binary,
});

pub fn init(
    interpreter: *Interpreter,
    shell_state: *ShellExecEnv,
    node: *const ast.Binary,
    parent: ParentPtr,
    io: IO,
) *Binary {
    var binary = parent.create(Binary);
    binary.node = node;
    binary.base = State.initWithNewAllocScope(.binary, interpreter, shell_state);
    binary.parent = parent;
    binary.io = io;
    binary.left = null;
    binary.right = null;
    binary.currently_executing = null;
    return binary;
}

pub fn start(this: *Binary) Yield {
    log("binary start {x} ({s})", .{ @intFromPtr(this), @tagName(this.node.op) });
    if (comptime bun.Environment.allow_assert) {
        assert(this.left == null);
        assert(this.right == null);
        assert(this.currently_executing == null);
    }

    this.currently_executing = this.makeChild(true);
    if (this.currently_executing == null) {
        this.currently_executing = this.makeChild(false);
        this.left = 0;
    }
    bun.assert(this.currently_executing != null);
    return this.currently_executing.?.start();
}

fn makeChild(this: *Binary, left: bool) ?ChildPtr {
    const node = if (left) &this.node.left else &this.node.right;
    switch (node.*) {
        .cmd => {
            const cmd = Cmd.init(this.base.interpreter, this.base.shell, node.cmd, Cmd.ParentPtr.init(this), this.io.copy());
            return ChildPtr.init(cmd);
        },
        .binary => {
            const binary = Binary.init(this.base.interpreter, this.base.shell, node.binary, Binary.ParentPtr.init(this), this.io.copy());
            return ChildPtr.init(binary);
        },
        .pipeline => {
            const pipeline = Pipeline.init(this.base.interpreter, this.base.shell, node.pipeline, Pipeline.ParentPtr.init(this), this.io.copy());
            return ChildPtr.init(pipeline);
        },
        .assign => |assigns| {
            const assign = Assigns.init(this.base.interpreter, this.base.shell, assigns, .shell, Assigns.ParentPtr.init(this), this.io.copy());
            return ChildPtr.init(assign);
        },
        .subshell => {
            switch (Subshell.initDupeShellState(
                this.base.interpreter,
                this.base.shell,
                node.subshell,
                Subshell.ParentPtr.init(this),
                this.io.copy(),
            )) {
                .result => |subshell| {
                    return ChildPtr.init(subshell);
                },
                .err => |e| {
                    this.base.throw(&bun.shell.ShellErr.newSys(e));
                    return null;
                },
            }
        },
        .@"if" => {
            const if_clause = If.init(this.base.interpreter, this.base.shell, node.@"if", If.ParentPtr.init(this), this.io.copy());
            return ChildPtr.init(if_clause);
        },
        .condexpr => {
            const condexpr = CondExpr.init(this.base.interpreter, this.base.shell, node.condexpr, CondExpr.ParentPtr.init(this), this.io.copy());
            return ChildPtr.init(condexpr);
        },
        .async => {
            const async = Async.init(this.base.interpreter, this.base.shell, node.async, Async.ParentPtr.init(this), this.io.copy());
            return ChildPtr.init(async);
        },
    }
}

pub fn childDone(this: *Binary, child: ChildPtr, exit_code: ExitCode) Yield {
    if (comptime bun.Environment.allow_assert) {
        assert(this.left == null or this.right == null);
        assert(this.currently_executing != null);
    }
    log("binary child done {x} ({s}) {s}", .{ @intFromPtr(this), @tagName(this.node.op), if (this.left == null) "left" else "right" });

    child.deinit();
    this.currently_executing = null;

    if (this.left == null) {
        this.left = exit_code;
        if ((this.node.op == .And and exit_code != 0) or (this.node.op == .Or and exit_code == 0)) {
            return this.parent.childDone(this, exit_code);
        }

        this.currently_executing = this.makeChild(false);
        if (this.currently_executing == null) {
            this.right = 0;
            return this.parent.childDone(this, 0);
        }

        return this.currently_executing.?.start();
    }

    this.right = exit_code;
    return this.parent.childDone(this, exit_code);
}

pub fn deinit(this: *Binary) void {
    if (this.currently_executing) |child| {
        child.deinit();
    }
    this.io.deinit();
    this.base.endScope();
    this.parent.allocator().destroy(this);
}

const bun = @import("bun");
const assert = bun.assert;

const ExitCode = bun.shell.ExitCode;
const Yield = bun.shell.Yield;
const ast = bun.shell.AST;

const Interpreter = bun.shell.Interpreter;
const Assigns = bun.shell.Interpreter.Assigns;
const Async = bun.shell.Interpreter.Async;
const Cmd = bun.shell.Interpreter.Cmd;
const CondExpr = bun.shell.Interpreter.CondExpr;
const IO = bun.shell.Interpreter.IO;
const If = bun.shell.Interpreter.If;
const Pipeline = bun.shell.Interpreter.Pipeline;
const ShellExecEnv = Interpreter.ShellExecEnv;
const State = bun.shell.Interpreter.State;
const Stmt = bun.shell.Interpreter.Stmt;
const Subshell = bun.shell.Interpreter.Subshell;

const StatePtrUnion = bun.shell.interpret.StatePtrUnion;
const log = bun.shell.interpret.log;

File: src/shell/states/Binary.rs
use bun_shell::interpret::{log, StatePtrUnion};
use bun_shell::interpreter::{
    Assigns, Async, Cmd, CondExpr, If, Interpreter, Pipeline, ShellExecEnv, State, Stmt, Subshell,
    IO,
};
use bun_shell::{ast, ExitCode, ShellErr, Yield};

pub struct Binary<'a> {
    pub base: State,
    // LIFETIMES.tsv: BORROW_PARAM — set from init param; deinit() never frees it
    pub node: &'a ast::Binary,
    /// Based on precedence rules binary expr can only be child of a stmt or
    /// another binary expr
    // LIFETIMES.tsv: BACKREF — parent.create allocs self; parent.allocator().destroy(this)
    pub parent: ParentPtr,
    pub left: Option<ExitCode>,
    pub right: Option<ExitCode>,
    pub io: IO,
    // LIFETIMES.tsv: OWNED — make_child allocs child; deinit() child.deinit() destroys
    pub currently_executing: Option<ChildPtr>,
}

pub type ChildPtr = StatePtrUnion<(
    Async,
    Cmd,
    Pipeline,
    Binary<'static>, // TODO(port): lifetime — self-referential variant in tagged ptr union
    Assigns,
    If,
    CondExpr,
    Subshell,
)>;

pub type ParentPtr = StatePtrUnion<(
    Stmt,
    Binary<'static>, // TODO(port): lifetime — self-referential variant in tagged ptr union
)>;

impl<'a> Binary<'a> {
    pub fn init(
        interpreter: &mut Interpreter,
        shell_state: &mut ShellExecEnv,
        node: &'a ast::Binary,
        parent: ParentPtr,
        io: IO,
    ) -> *mut Binary<'a> {
        // TODO(port): parent.create(Binary) allocates `Binary` from the parent's allocator
        // and returns an uninitialized *mut Binary. Phase B: decide whether StatePtrUnion::create
        // returns &mut MaybeUninit<Self> or Box<Self>.
        let binary: &mut Binary<'a> = parent.create::<Binary<'a>>();
        binary.node = node;
        binary.base = State::init_with_new_alloc_scope(StateKind::Binary, interpreter, shell_state);
        binary.parent = parent;
        binary.io = io;
        binary.left = None;
        binary.right = None;
        binary.currently_executing = None;
        binary
    }

    pub fn start(&mut self) -> Yield {
        log!(
            "binary start {:x} ({})",
            self as *const _ as usize,
            <&'static str>::from(self.node.op)
        );
        debug_assert!(self.left.is_none());
        debug_assert!(self.right.is_none());
        debug_assert!(self.currently_executing.is_none());

        self.currently_executing = self.make_child(true);
        if self.currently_executing.is_none() {
            self.currently_executing = self.make_child(false);
            self.left = Some(0);
        }
        debug_assert!(self.currently_executing.is_some());
        self.currently_executing.as_mut().unwrap().start()
    }

    fn make_child(&mut self, left: bool) -> Option<ChildPtr> {
        let node = if left { &self.node.left } else { &self.node.right };
        match node {
            ast::Expr::Cmd(cmd_node) => {
                let cmd = Cmd::init(
                    self.base.interpreter,
                    self.base.shell,
                    cmd_node,
                    Cmd::ParentPtr::init(self),
                    self.io.copy(),
                );
                Some(ChildPtr::init(cmd))
            }
            ast::Expr::Binary(binary_node) => {
                let binary = Binary::init(
                    self.base.interpreter,
                    self.base.shell,
                    binary_node,
                    Binary::ParentPtr::init(self),
                    self.io.copy(),
                );
                Some(ChildPtr::init(binary))
            }
            ast::Expr::Pipeline(pipeline_node) => {
                let pipeline = Pipeline::init(
                    self.base.interpreter,
                    self.base.shell,
                    pipeline_node,
                    Pipeline::ParentPtr::init(self),
                    self.io.copy(),
                );
                Some(ChildPtr::init(pipeline))
            }
            ast::Expr::Assign(assigns) => {
                let assign = Assigns::init(
                    self.base.interpreter,
                    self.base.shell,
                    assigns,
                    AssignCtx::Shell,
                    Assigns::ParentPtr::init(self),
                    self.io.copy(),
                );
                Some(ChildPtr::init(assign))
            }
            ast::Expr::Subshell(subshell_node) => {
                match Subshell::init_dupe_shell_state(
                    self.base.interpreter,
                    self.base.shell,
                    subshell_node,
                    Subshell::ParentPtr::init(self),
                    self.io.copy(),
                ) {
                    Ok(subshell) => Some(ChildPtr::init(subshell)),
                    Err(e) => {
                        self.base.throw(&ShellErr::new_sys(e));
                        None
                    }
                }
            }
            ast::Expr::If(if_node) => {
                let if_clause = If::init(
                    self.base.interpreter,
                    self.base.shell,
                    if_node,
                    If::ParentPtr::init(self),
                    self.io.copy(),
                );
                Some(ChildPtr::init(if_clause))
            }
            ast::Expr::Condexpr(condexpr_node) => {
                let condexpr = CondExpr::init(
                    self.base.interpreter,
                    self.base.shell,
                    condexpr_node,
                    CondExpr::ParentPtr::init(self),
                    self.io.copy(),
                );
                Some(ChildPtr::init(condexpr))
            }
            ast::Expr::Async(async_node) => {
                let r#async = Async::init(
                    self.base.interpreter,
                    self.base.shell,
                    async_node,
                    Async::ParentPtr::init(self),
                    self.io.copy(),
                );
                Some(ChildPtr::init(r#async))
            }
        }
    }

    pub fn child_done(&mut self, child: ChildPtr, exit_code: ExitCode) -> Yield {
        debug_assert!(self.left.is_none() || self.right.is_none());
        debug_assert!(self.currently_executing.is_some());
        log!(
            "binary child done {:x} ({}) {}",
            self as *const _ as usize,
            <&'static str>::from(self.node.op),
            if self.left.is_none() { "left" } else { "right" }
        );

        child.deinit();
        self.currently_executing = None;

        if self.left.is_none() {
            self.left = Some(exit_code);
            if (self.node.op == ast::BinaryOp::And && exit_code != 0)
                || (self.node.op == ast::BinaryOp::Or && exit_code == 0)
            {
                return self.parent.child_done(self, exit_code);
            }

            self.currently_executing = self.make_child(false);
            if self.currently_executing.is_none() {
                self.right = Some(0);
                return self.parent.child_done(self, 0);
            }

            return self.currently_executing.as_mut().unwrap().start();
        }

        self.right = Some(exit_code);
        self.parent.child_done(self, exit_code)
    }

    // TODO(port): not `impl Drop` — this is a self-destroying deinit that frees `self` via the
    // parent's allocator (parent-owned arena allocation). Drop cannot take params or destroy
    // its own backing storage. Phase B: reconcile with StatePtrUnion ownership model.
    pub fn deinit(&mut self) {
        if let Some(child) = self.currently_executing.take() {
            child.deinit();
        }
        self.io.deinit();
        self.base.end_scope();
        // SAFETY: `self` was allocated by `parent.create(Binary)` from this allocator;
        // after this call `self` is freed and must not be accessed.
        unsafe {
            self.parent.allocator().destroy(self);
        }
    }
}

// TODO(port): `StateKind::Binary` / `AssignCtx::Shell` — exact enum names from sibling modules
use bun_shell::interpreter::{AssignCtx, StateKind};

// ──────────────────────────────────────────────────────────────────────────
// PORT STATUS
//   source:     src/shell/states/Binary.zig (174 lines)
//   confidence: medium
//   todos:      4
//   notes:      StatePtrUnion self-ref lifetime + parent-allocator self-destroy need Phase B ownership design
// ──────────────────────────────────────────────────────────────────────────
